### DISCUSSION TITLE
Default exchange missing after upgrade rabbit

### DISCUSSION THREAD

**Discussion by thong234** [rabbitmq/rabbitmq-server#10363] [Other]
Hi, 

We're trying to upgrade rabbitmq from rabbit 3.12.2 to 3.12.12,
After update everything seems to be ok except that default exchange such as (AMQP default),[amq.direct],[amq.fanout]  seems to be missing from every vhost.
Are there anyway I can recover those exchange as I can't seems to create it from rabbitmq ui/

**Comment by michaelklishin** [comment]
They are created at virtual host creation time and never deleted by RabbitMQ itself. There is no way to create them other than a moderately complex expression fed to `rabbitmqctl eval`, since they are never removed.

Here's a code snippet from `rabbit_vhost` that sets them up:

``` erl
_ = [begin
     Resource = rabbit_misc:r(Name, exchange, ExchangeName),
     rabbit_log:debug("Will declare an exchange ~tp", [Resource]),
     _ = rabbit_exchange:declare(Resource, Type, true, false, Internal, [], ActingUser)
 end || {ExchangeName, Type, Internal} <-
        [{<<"amq.direct">>,         direct,  false},
         {<<"amq.topic">>,          topic,   false},
         %% per 0-9-1 pdf
         {<<"amq.match">>,          headers, false},
         %% per 0-9-1 xml
         {<<"amq.headers">>,        headers, false},
         {<<"amq.fanout">>,         fanout,  false},
         {<<"amq.rabbitmq.trace">>, topic,   true}]].
```

The default exchange (the one with the blank name) is not actually an exchange in 3.12 with all feature flags enabled, it is a special path in the code.

I'm afraid we cannot suggest much with this amount of information.
